macro-rules/src/util.rs

```diff
@@ -0,0 +1,25 @@
+#[macro_export]
+macro_rules! macro_if {
+    (if [] { $($then:tt)* }) => {};
+
+    (if false { $($then:tt)* }) => {};
+
+    (if [$($input:tt)+] { $($then:tt)* }) => {
+        $($then)*
+    };
+
+    (if true { $($then:tt)* }) => {
+        $($then)*
+    };
+
+    // Testing what kind of type we have (scalar vs ref)
+    //
+    // see java_types.rs
+
+    (if is_ref_ty($t:ident) { $($then:tt)* }) => {};
+
+    (if is_ref_ty($t:tt) { $($then:tt)* }) => {
```

Comment:
this works because idents are never references but token trees are?

Reply:
Yes, it's part of the invariant for how we represent types -- an identifier is a scalar or void, a `()` grouping is a reference. I could replace `$t:tt` with `($(t:tt)*)`, that might be better.